### DISCUSSION TITLE
Titel und Untertitel als TradeLineItem

### DISCUSSION THREAD

**Discussion by mightyScoop** [stephanstapel/ZUGFeRD-csharp#410] [General]
Hallo Stephan,

zunächst einmal vielen Dank für Dein Projekt – Respekt für Deine großartige Arbeit!

Nun zu meiner Frage: Ab XRechnung 3.x ist es möglich, in den Positionen eine Titel-/Untertitel-Struktur abzubilden, die jeweils über eigene Zu- und Abschläge verfügt. Wird diese Funktionalität auch in Deinem Projekt unterstützt?

Viele Grüße

Dominic

**Comment by stephanstapel** [comment]
Hallo @mightyScoop ,
meinst Du dies:
https://github.com/stephanstapel/ZUGFeRD-csharp/pull/404
?

**Reply by mightyScoop** [reply]
Das sieht tatsächlich sehr gut aus. Werde ich am Montag testen.
Vielen Dank!

**Reply by HenriSHS** [reply]
Hallo zusammen,
wollte nur Anmerken das ich die IncludedReferencedProduct-Funktion aus #404 implementiert habe, um Stücklistenpositionen abbilden zu können. Zumal man laut BG-X-1 auch keine Price Details implementieren kann, bin ich mir nicht ganz sicher ob diese Funktion zu dem Problem passt. Meiner Auffassung nacht könnte man BT-X-304 implementieren, um eine Hierachy zu schaffen.

![grafik](https://github.com/user-attachments/assets/595b3588-6741-4a18-8aa9-3f2a5d2d2f2c)

**Comment by mightyScoop** [comment]
Hallo zusammen,

ja, ich glaube auch nicht, dass mit BG-X-1 eine Hierarchie abgebildet werden kann. 
Ist denn Factur-X BT-X-304 äquivalent mit
XRechnung BG-DEX-01?

![image](https://github.com/user-attachments/assets/4ec91aeb-706c-464b-89ec-1245061041d8)


Grüße

Dominic

**Reply by HenriSHS** [reply]
> Ist denn Factur-X BT-X-304 äquivalent mit XRechnung BG-DEX-01?

Ich würde sagen ja

**Reply by mightyScoop** [reply]
Hast Du vor auch BT-X-304 zu implementieren? Besteht die Hoffnung?

Grüße
Dominic

**Reply by HenriSHS** [reply]
Ich habe nicht geplant BT-X-304 zu implementieren, da ich selbst bisher keine Verwendung dafür hätte. Vielleicht ist es eine Option wenn du selbst die Funktion implementierst? Ich weiß das @stephanstapel dir gerne bei deiner Pull Request weiterhilft, um Fehler zu beheben und den Code zu verbessern.

**Reply by HenriSHS** [reply]
@mightyScoop hab meine Meinung geändert ich arbeite an der Implementierung. #430

**Reply by mightyScoop** [reply]
Oh großartig,

das sind super Neuigkeiten. Vielen Dank schon einmal!

**Reply by HenriSHS** [reply]
@mightyScoop Die Funktion kann ab jetzt verwendet werden. Außerdem könnte #426 für dich interssant sein. 

> BT-X-7 - Indicating whether an item includes the prices which must be taken into account when calculating the invoice amount, or whether it only contains information.

**Reply by mightyScoop** [reply]
Hallo HenriSHS,

erstmal Danke für die Implementierung! 

Habe noch eine Frage.
Folgender Fall:
2 Positionen (sagen wir mit je 100€) sind zusammengefasst unter einem Titel.
D.h. der Titel hat jetzt 200€ Gesamt. Auf den Titel gibt es einen Zuschlag von 10€. Wie würdest Du das abbilden?
In Deinen Beispielen hast Du ja den UnitPrice der Titel auf 0 gesetzt.
Dafür brauchen wir wahrscheinlich  SetLineStatus(), oder?

Grüße

Dominic

**Reply by HenriSHS** [reply]
Meiner Meinung nach gibt es verschiedene Ansätze, um den beschriebenen Fall korrekt darzustellen. Ich will dazu sagen ich bin kein Profi, aber ich würde es vermutlich wie folgt lösen. Hier sind drei mögliche Lösungswege, die mir einfallen:

1. Titel als eigene Position mit Unterpositionen    
Man könnte eine Position für den Titel erstellen und die beiden anderen Positionen als IncludedReferencedProduct hinzufügen. Das würde bedeuten, dass du bei den beiden Unterpositionen nur den Titel und die Anzahl angibst. Den Zuschlag könntest du dann direkt bei der Titelposition hinzufügen.

```
            TradeLineItem tradeLineItem = desc.AddTradeLineItem(
                lineID: "1",
                name: "Titel (Position 1)",
                billedQuantity: 1m,
                unitCode: QuantityCodes.H87,
                netUnitPrice: 200m,
                categoryCode: TaxCategoryCodes.S,
                taxPercent: 19.0m,
                taxType: TaxTypes.VAT);
            tradeLineItem.AddTradeAllowanceCharge(false, CurrencyCodes.EUR, 200, 10, "Zuschlag für Titel");
            tradeLineItem.AddIncludedReferencedProduct("Unterposition 1", 1, QuantityCodes.C62);
            tradeLineItem.AddIncludedReferencedProduct("Unterposition 2", 1, QuantityCodes.C62);
```

2. Drei eigenständige Positionen mit Verknüpfungen
Eine andere Möglichkeit wäre, alle drei Positionen normal zu erstellen. Dann würdest du die Unterpositionen der Titelposition zuordnen. In diesem Fall könntest du den Unterpositionen einen LineStatus setzen, damit ihre Beträge nicht zum Gesamtbetrag addiert werden. Den Gesamtbetrag würdest du in der übergeordneten Titelposition abbilden, inklusive des Zuschlags.
```
            TradeLineItem tradeLineItem = desc.AddTradeLineItem(
                lineID: "1",
                name: "Titel (Position 1)",
                billedQuantity: 1m,
                unitCode: QuantityCodes.H87,
                netUnitPrice: 200m,
                categoryCode: TaxCategoryCodes.S,
                taxPercent: 19.0m,
                taxType: TaxTypes.VAT);
            tradeLineItem.SetLineStatus(LineStatusCodes.DocumentationClaim, LineStatusReasonCodes.INFORMATION); //Line Status setzen damit PosBetrag nicht zu Gesamtbetrag zählt

            TradeLineItem subTradeLineItem1 = desc.AddTradeLineItem(
                lineID: "1.1",
                name: "Unterposition 1",
                billedQuantity: 1,
                unitCode: QuantityCodes.C62,
                netUnitPrice: 100,
                categoryCode: TaxCategoryCodes.S,
                taxPercent: 19.0m,
                taxType: TaxTypes.VAT);
            subTradeLineItem1.SetParentLineId("1"); //Postion als Übergeordnetes Element setzen
            tradeLineItem.AddTradeAllowanceCharge(false, CurrencyCodes.EUR, 100, 5, "Zuschlag für Unterposition 1");

            TradeLineItem subTradeLineItem2 = desc.AddTradeLineItem(
                lineID: "1.2",
                name: "Unterposition 2",
                billedQuantity: 1,
                unitCode: QuantityCodes.C62,
                netUnitPrice: 100,
                categoryCode: TaxCategoryCodes.S,
                taxPercent: 19.0m,
                taxType: TaxTypes.VAT);
            subTradeLineItem1.SetParentLineId("1"); //Postion als Übergeordnetes Element setzen
            tradeLineItem.AddTradeAllowanceCharge(false, CurrencyCodes.EUR, 100, 5, "Zuschlag für Unterposition 2");
```


3. Mein wahrscheinlich favorisierter Ansatz für diesen Fall
Du erstellst eine übergeordnete Position für den Titel und fügst die Unterpositionen hinzu. Den Zuschlag würdest du der übergeordneten Position hinzufügen, während die Unterpositionen einen LineStatus bekommen, dass sie nur der Dokumentation dienen und nicht zum Gesamtbetrag zählen.

```
TradeLineItem tradeLineItem = desc.AddTradeLineItem(
    lineID: "1",
    name: "Titel (Position 1)",
    billedQuantity: 1m,
    unitCode: QuantityCodes.H87,
    netUnitPrice: 200m,
    categoryCode: TaxCategoryCodes.S,
    taxPercent: 19.0m,
    taxType: TaxTypes.VAT);
tradeLineItem.AddTradeAllowanceCharge(false, CurrencyCodes.EUR, 200, 10, "Zuschlag für Titel");

TradeLineItem subTradeLineItem1 = desc.AddTradeLineItem(
    lineID: "1.1",
    name: "Unterposition 1",
    billedQuantity: 1,
    unitCode: QuantityCodes.C62,
    netUnitPrice: 100,
    categoryCode: TaxCategoryCodes.S,
    taxPercent: 19.0m,
    taxType: TaxTypes.VAT);
subTradeLineItem1.SetLineStatus(LineStatusCodes.DocumentationClaim, LineStatusReasonCodes.INFORMATION);
subTradeLineItem1.SetParentLineId("1");

TradeLineItem subTradeLineItem2 = desc.AddTradeLineItem(
    lineID: "1.2",
    name: "Unterposition 2",
    billedQuantity: 1,
    unitCode: QuantityCodes.C62,
    netUnitPrice: 100,
    categoryCode: TaxCategoryCodes.S,
    taxPercent: 19.0m,
    taxType: TaxTypes.VAT);
subTradeLineItem2.SetLineStatus(LineStatusCodes.DocumentationClaim, LineStatusReasonCodes.INFORMATION);
subTradeLineItem2.SetParentLineId("1");
```

**Reply by mightyScoop** [reply]
Hallo Henri,

vielen Dank für die drei Lösungsansätze!
Ich denke, Ansatz Nr. 3 ist der vielversprechendste Weg.

Bei Nr. 1 fehlen tatsächlich zu viele Informationen zu den Sub-Positionen. Eine Referenz reicht für meine Bedürfnisse (leider) nicht aus.

Ansatz Nr. 2 könnte in einfachen Fällen zwar ebenfalls funktionieren, jedoch ergeben sich Probleme, wenn:

Zu- und Abschläge nicht pauschal, sondern prozentual auf die Titel berechnet werden, und
die Sub-Positionen Mengen ungleich 1 aufweisen.
In solchen Szenarien ist eine Umrechnung der Zu-/Abschläge auf die Sub-Positionen nicht mehr wirklich praktikabel.

Nr. 3 klingt hingegen hervorragend. Genau so hätte ich es vermutlich auch versucht!
Funktioniert dieser Ansatz auch bei einer noch größeren Hierarchie? Zum Beispiel bei einer Struktur wie Titel -> Untertitel -> Unter-Untertitel -> Positionen?
Ich konnte das leider noch nicht ausprobieren, da mir der Code für SetLineStatus() noch nicht vorliegt.

Viele Grüße
Dominic

**Reply by HenriSHS** [reply]
Freut mich das der Ansatz für dich passt. Eigentlich sollte damit auch beliebig große Hierarchien erstellt werden können.
